### ISSUE TITLE
Use pb file in python with tensorflow to get AgentActions

### ISSUE THREAD

**Issue by SaitaoGit** [Unity-Technologies/ml-agents#2077]
I use ML-Agents in unity to train a simple linefollow robot. For that i use visual observation with one camera and vector actions with space type discrete (1 branch with 4 options). The training works and i got my nn file as well as the matching pb file. Now i want to use that pb file in a python script to run the same prediction outside of unity. I'm completely new to tensorflow and do not know how i can reproduce my result that i get in float[] vectorAction (values between 0 and 3).

The code I use to start a session:
  ```
     with tf.Session() as sess:
         print("load graph")
         with tf.gfile.GFile(GRAPH_PB_PATH, 'rb') as f:
             graph_def = tf.GraphDef()
         graph_def.ParseFromString(f.read())
         sess.graph.as_default()
         tf.import_graph_def(graph_def, name='')
         graph_nodes = [n for n in graph_def.node]
         names = []
         for t in graph_nodes:
             names.append(t.name)
         print(names)
         logits = sess.graph.get_tensor_by_name("visual_observation_0:0")
```
The code I use for a prediction:
```
 cv2_edges = numpy.array(cv2_edges).reshape(1, 30, 80, 1)
 prediciton = self.session.run(self.logits, {logits: cv2_edges})
```

When I print my prediction I get this as a result:

```
 Prediction: [[[ 64.]
   [  0.]
   [  0.]
   ...
   [  0.]
   [  0.]
   [  0.]]
 
  [[128.]
   [  0.]
   [  0.]
   ...
   [  0.]
   [  0.]
   [  0.]]
 
  [[128.]
   [  0.]
   [  0.]
   ...
   [  0.]
   [  0.]
   [  0.]]
 
  ...
 
  [[  0.]
   [  0.]
   [  0.]
   ...
   [  0.]
   [  0.]
   [  0.]]
 
  [[  0.]
   [  0.]
   [  0.]
   ...
   [  0.]
   [  0.]
   [  0.]]
 
  [[  0.]
   [  0.]
   [  0.]
   ...
   [  0.]
   [  0.]
   [  0.]]]
```
Maybe it's a stupid question, but i was expection a float[] where I can perform a similar operation like Mathf.FloorToInt() in order to get my discrete action. Do you know how i can get to my expected result?

**Comment by vincentpierre** [comment]
I would recommend looking at the TensorFlow project for help. This seems to be outside the scope of ML-Agents. The output of the network (the `action` tensor) should be log probabilities for each action, you would then need to apply the multinomial operator to sample the actions from that distribution.  I am not sure what `self.logits` is but what you obtain seems to not have the expected shape nor the right values.

**Comment by SaitaoGit** [comment]
Thanks for your help! The problem was that i was using the wrong tensor... 
When I use my trained model in a real robot, do I still need to apply the multinominal operator? As far as I understand the multinominal operator is used for exploration during training ( so that the agent will try new options even if he found a good solution to the current state in order to possibly achieve an even better solution), but when I just use the trained model to make my real robot follow a line I don't need exploration anymore, do I? For that I just need the next action with the highest probability?

**Comment by vincentpierre** [comment]
I think both approaches are valid, I would use multinomial still just to make sure the robot does not stuck itself in a loop. If two actions have very close probabilities, it does not really make sense to me to always pick the highest one...
Glad you got it to work.

**Comment by xiaomaogy** [comment]
Thank you for the discussion. We are closing this issue due to inactivity. Feel free to reopen it if you’d like to continue the discussion though.
